File: apps/TextBoomPpOcrBench/src/com/smartisax/ocrbench/MainActivity.java
package com.smartisax.ocrbench;

import android.app.Activity;
import android.content.res.AssetManager;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.os.Bundle;
import android.os.Debug;
import android.os.SystemClock;
import android.util.Log;
import android.view.Gravity;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

public final class MainActivity extends Activity {
    private static final String TAG = "SmartisaxOcrBench";
    private static final String EXTRA_IMAGE_PATH = "image_path";
    private static final String EXTRA_RESULT_PATH = "result_path";
    private static final String DEFAULT_IMAGE_NAME = "imageboom.jpg";
    private static final String MODEL_ASSET_ROOT = "ppocr-v2";
    private static final String MODEL_ID = "ch_ppocr_mobile_v2.0_slim_opt_for_cpu_v2_10_rc";
    private static final String DET_MODEL_ASSET = MODEL_ASSET_ROOT + "/models/ch_ppocr_mobile_v2.0_det_slim_opt.nb";
    private static final String CLS_MODEL_ASSET = MODEL_ASSET_ROOT + "/models/ch_ppocr_mobile_v2.0_cls_slim_opt.nb";
    private static final String REC_MODEL_ASSET = MODEL_ASSET_ROOT + "/models/ch_ppocr_mobile_v2.0_rec_slim_opt.nb";
    private static final String CONFIG_ASSET = MODEL_ASSET_ROOT + "/config.txt";
    private static final String LABEL_ASSET = MODEL_ASSET_ROOT + "/labels/ppocr_keys_v1.txt";
    private static final String CPU_POWER_MODE = "LITE_POWER_HIGH";
    private static final int CPU_THREAD_NUM = 4;
    private static final boolean NATIVE_READY;
    private static final String NATIVE_LOAD_ERROR;

    static {
        boolean ready = false;
        String error = null;
        try {
            System.loadLibrary("c++_shared");
            System.loadLibrary("paddle_light_api_shared");
            System.loadLibrary("smartisax_ppocr_bench");
            ready = true;
        } catch (Throwable throwable) {
            error = throwable.getClass().getSimpleName() + ": " + throwable.getMessage();
            Log.e(TAG, "native OCR runtime load failed", throwable);
        }
        NATIVE_READY = ready;
        NATIVE_LOAD_ERROR = error;
    }

    private static native String nativeRunPpOcr(
            String imagePath,
            String detModelPath,
            String clsModelPath,
            String recModelPath,
            String configPath,
            String labelPath,
            int cpuThreadNum,
            String cpuPowerMode
    );

    private TextView statusView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(createContentView());
        statusView.setText("Preparing OCR benchmark...");

        new Thread(new Runnable() {
            @Override
            public void run() {
                final String message = runBenchmark();
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        statusView.setText(message);
                    }
                });
            }
        }, "smartisax-ocrbench").start();
    }

    private ScrollView createContentView() {
        ScrollView scrollView = new ScrollView(this);
        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setGravity(Gravity.START);
        int padding = dp(20);
        layout.setPadding(padding, padding, padding, padding);

        TextView title = new TextView(this);
        title.setText("Smartisax OCR Bench");
        title.setTextSize(22);
        title.setGravity(Gravity.START);
        layout.addView(title);

        statusView = new TextView(this);
        statusView.setTextSize(14);
        statusView.setPadding(0, dp(14), 0, 0);
        statusView.setTextIsSelectable(true);
        layout.addView(statusView);

        scrollView.addView(layout);
        return scrollView;
    }

    private int dp(int value) {
        return Math.round(value * getResources().getDisplayMetrics().density);
    }

    private String runBenchmark() {
        File appRoot = getExternalFilesDir(null);
        if (appRoot == null) {
            appRoot = getFilesDir();
        }

        File image = resolvePath(
                getIntent().getStringExtra(EXTRA_IMAGE_PATH),
                new File(new File(appRoot, "input"), DEFAULT_IMAGE_NAME)
        );
        File result = resolvePath(
                getIntent().getStringExtra(EXTRA_RESULT_PATH),
                new File(new File(appRoot, "results"), "last-result.json")
        );

        try {
            JSONObject payload = buildPredictionPayload(image);
            writeJson(result, payload);
            String message = "result=" + payload.optString("result") + "\n"
                    + "image=" + image.getAbsolutePath() + "\n"
                    + "output=" + result.getAbsolutePath();
            Log.i(TAG, message);
            return message;
        } catch (Exception error) {
            Log.e(TAG, "benchmark failed", error);
            return "result=ERROR\n" + error.getClass().getSimpleName() + ": " + error.getMessage();
        }
    }

    private File resolvePath(String path, File fallback) {
        if (path == null || path.trim().isEmpty()) {
            return fallback;
        }
        File file = new File(path.trim());
        if (file.isAbsolute()) {
            return file;
        }
        File appRoot = getExternalFilesDir(null);
        if (appRoot == null) {
            appRoot = getFilesDir();
        }
        return new File(appRoot, path.trim());
    }

    private JSONObject buildPredictionPayload(File image) throws JSONException, IOException {
        long started = SystemClock.elapsedRealtime();
        ImageProbe probe = probeImage(image);
        JSONObject nativeResult = runNativeIfReady(image, probe);
        long elapsedMs = SystemClock.elapsedRealtime() - started;
        JSONArray ppocr = nativeResult.optJSONArray("lines");
        if (ppocr == null) {
            ppocr = new JSONArray();
        }
        String status = nativeResult.optString("status", probe.status);
        double nativeTotalMs = nativeResult.optDouble("total_ms", elapsedMs);

        JSONObject sample = new JSONObject();
        sample.put("id", imageId(image));
        sample.put("source_image", image.getAbsolutePath());
        sample.put("status", status);
        sample.put("image_sha256", probe.sha256 == null ? JSONObject.NULL : probe.sha256);
        sample.put("image_size", imageSizeArray(
                nativeResult.optInt("image_width", probe.width),
                nativeResult.optInt("image_height", probe.height)
        ));
        sample.put("latency_ms", Math.round(nativeTotalMs));
        sample.put("peak_pss_kb", currentPssKb());
        sample.put("ppocr", ppocr);
        sample.put("native_metrics", nativeMetrics(nativeResult));
        if (probe.error != null) {
            sample.put("error", probe.error);
        }
        if (nativeResult.has("error")) {
            sample.put("native_error", nativeResult.optString("error"));
        }

        JSONObject engine = new JSONObject();
        engine.put("id", "smartisax-paddle-lite-ppocr-v2-mobile-slim");
        engine.put("version", "0.2.0");
        engine.put("runtime", "android-native-paddle-lite");
        engine.put("status", NATIVE_READY ? "native-ready" : "native-load-error");
        if (NATIVE_LOAD_ERROR != null) {
            engine.put("load_error", NATIVE_LOAD_ERROR);
        }

        JSONObject model = new JSONObject();
        model.put("id", MODEL_ID);
        model.put("family", "PP-OCR mobile v2 slim");
        model.put("det", DET_MODEL_ASSET);
        model.put("cls", CLS_MODEL_ASSET);
        model.put("rec", REC_MODEL_ASSET);
        model.put("labels", LABEL_ASSET);
        model.put("config", CONFIG_ASSET);
        model.put("cpu_threads", CPU_THREAD_NUM);
        model.put("cpu_power_mode", CPU_POWER_MODE);

        JSONObject device = new JSONObject();
        device.put("manufacturer", Build.MANUFACTURER);
        device.put("brand", Build.BRAND);
        device.put("model", Build.MODEL);
        device.put("device", Build.DEVICE);
        device.put("sdk_int", Build.VERSION.SDK_INT);

        JSONArray samples = new JSONArray();
        samples.put(sample);

        JSONObject payload = new JSONObject();
        payload.put("generated_at", utcNow());
        payload.put("kind", "textboom-ppocr-runtime-prediction");
        payload.put("boundary", "standalone benchmark APK; no TextBoom, ROM, or data mutation");
        payload.put("engine", engine);
        payload.put("model", model);
        payload.put("device", device);
        payload.put("samples", samples);
        payload.put("result", status);
        return payload;
    }

    private JSONObject runNativeIfReady(File image, ImageProbe probe) throws JSONException, IOException {
        if (!"IMAGE_READY".equals(probe.status)) {
            JSONObject result = new JSONObject();
            result.put("status", probe.status);
            if (probe.error != null) {
                result.put("error", probe.error);
            }
            return result;
        }
        if (!NATIVE_READY) {
            JSONObject result = new JSONObject();
            result.put("status", "NATIVE_LOAD_ERROR");
            result.put("error", NATIVE_LOAD_ERROR == null ? "unknown native load error" : NATIVE_LOAD_ERROR);
            return result;
        }

        ModelFiles model = ensureModelFiles();
        String raw = nativeRunPpOcr(
                image.getAbsolutePath(),
                model.det.getAbsolutePath(),
                model.cls.getAbsolutePath(),
                model.rec.getAbsolutePath(),
                model.config.getAbsolutePath(),
                model.labels.getAbsolutePath(),
                CPU_THREAD_NUM,
                CPU_POWER_MODE
        );
        return new JSONObject(raw);
    }

    private JSONObject nativeMetrics(JSONObject nativeResult) throws JSONException {
        JSONObject metrics = new JSONObject();
        metrics.put("det_ms", nativeResult.optDouble("det_ms", 0.0));
        metrics.put("rec_ms", nativeResult.optDouble("rec_ms", 0.0));
        metrics.put("total_ms", nativeResult.optDouble("total_ms", 0.0));
        metrics.put("line_count", nativeResult.optJSONArray("lines") == null ? 0 : nativeResult.optJSONArray("lines").length());
        return metrics;
    }

    private JSONArray imageSizeArray(int width, int height) {
        JSONArray imageSize = new JSONArray();
        imageSize.put(width);
        imageSize.put(height);
        return imageSize;
    }

    private ModelFiles ensureModelFiles() throws IOException {
        File root = new File(getFilesDir(), MODEL_ASSET_ROOT);
        File det = ensureAssetFile(DET_MODEL_ASSET, new File(root, "models/ch_ppocr_mobile_v2.0_det_slim_opt.nb"));
        File cls = ensureAssetFile(CLS_MODEL_ASSET, new File(root, "models/ch_ppocr_mobile_v2.0_cls_slim_opt.nb"));
        File rec = ensureAssetFile(REC_MODEL_ASSET, new File(root, "models/ch_ppocr_mobile_v2.0_rec_slim_opt.nb"));
        File config = ensureAssetFile(CONFIG_ASSET, new File(root, "config.txt"));
        File labels = ensureAssetFile(LABEL_ASSET, new File(root, "labels/ppocr_keys_v1.txt"));
        return new ModelFiles(det, cls, rec, config, labels);
    }

    private File ensureAssetFile(String assetPath, File target) throws IOException {
        if (target.isFile() && target.length() > 0) {
            return target;
        }
        File parent = target.getParentFile();
        if (parent != null && !parent.isDirectory() && !parent.mkdirs()) {
            throw new IOException("failed to create " + parent);
        }
        AssetManager assets = getAssets();
        InputStream input = assets.open(assetPath);
        try {
            OutputStream output = new FileOutputStream(target, false);
            try {
                byte[] buffer = new byte[16384];
                while (true) {
                    int read = input.read(buffer);
                    if (read < 0) {
                        break;
                    }
                    output.write(buffer, 0, read);
                }
            } finally {
                output.close();
            }
        } finally {
            input.close();
        }
        return target;
    }

    private ImageProbe probeImage(File image) {
        if (!image.isFile()) {
            return ImageProbe.error("IMAGE_MISSING", 0, 0, null, "missing file");
        }
        try {
            BitmapFactory.Options options = new BitmapFactory.Options();
            options.inJustDecodeBounds = true;
            BitmapFactory.decodeFile(image.getAbsolutePath(), options);
            if (options.outWidth <= 0 || options.outHeight <= 0) {
                return ImageProbe.error("IMAGE_DECODE_FAILED", 0, 0, sha256(image), "BitmapFactory returned empty bounds");
            }
            return new ImageProbe("IMAGE_READY", options.outWidth, options.outHeight, sha256(image), null);
        } catch (Exception error) {
            return ImageProbe.error("IMAGE_PROBE_ERROR", 0, 0, null, error.getClass().getSimpleName() + ": " + error.getMessage());
        }
    }

    private int currentPssKb() {
        Debug.MemoryInfo memoryInfo = new Debug.MemoryInfo();
        Debug.getMemoryInfo(memoryInfo);
        return memoryInfo.getTotalPss();
    }

    private String imageId(File image) {
        String name = image.getName();
        int dot = name.lastIndexOf('.');
        if (dot > 0) {
            return name.substring(0, dot);
        }
        return name.isEmpty() ? "sample" : name;
    }

    private void writeJson(File result, JSONObject payload) throws IOException {
        File parent = result.getParentFile();
        if (parent != null && !parent.isDirectory() && !parent.mkdirs()) {
            throw new IOException("failed to create " + parent);
        }
        FileWriter writer = new FileWriter(result, false);
        try {
            writer.write(payload.toString(2));
            writer.write("\n");
        } catch (JSONException error) {
            throw new IOException("failed to serialize JSON", error);
        } finally {
            writer.close();
        }
    }

    private String sha256(File file) throws IOException, NoSuchAlgorithmException {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        FileInputStream input = new FileInputStream(file);
        try {
            byte[] buffer = new byte[8192];
            while (true) {
                int read = input.read(buffer);
                if (read < 0) {
                    break;
                }
                digest.update(buffer, 0, read);
            }
        } finally {
            input.close();
        }
        byte[] bytes = digest.digest();
        StringBuilder builder = new StringBuilder(bytes.length * 2);
        for (byte value : bytes) {
            builder.append(String.format(Locale.US, "%02x", value & 0xff));
        }
        return builder.toString();
    }

    private String utcNow() {
        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
        format.setTimeZone(TimeZone.getTimeZone("UTC"));
        return format.format(new Date());
    }

    private static final class ModelFiles {
        final File det;
        final File cls;
        final File rec;
        final File config;
        final File labels;

        ModelFiles(File det, File cls, File rec, File config, File labels) {
            this.det = det;
            this.cls = cls;
            this.rec = rec;
            this.config = config;
            this.labels = labels;
        }
    }

    private static final class ImageProbe {
        final String status;
        final int width;
        final int height;
        final String sha256;
        final String error;

        ImageProbe(String status, int width, int height, String sha256, String error) {
            this.status = status;
            this.width = width;
            this.height = height;
            this.sha256 = sha256;
            this.error = error;
        }

        static ImageProbe error(String status, int width, int height, String sha256, String error) {
            return new ImageProbe(status, width, height, sha256, error);
        }
    }
}

File: apps/TextBoomPpOcrOfficialBench/src/main/java/com/smartisax/ocrbench/officialbench/MainActivity.kt
package com.smartisax.ocrbench.officialbench

import android.app.Activity
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Bundle
import android.os.Debug
import android.os.SystemClock
import android.util.Log
import android.view.Gravity
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import com.paddle.ocr.EngineConfig
import com.paddle.ocr.PaddleOCR
import com.paddle.ocr.PaddleOCRConfig
import com.paddle.ocr.model.OCRBox
import com.paddle.ocr.model.OCRRunResult
import com.paddle.ocr.util.OpenCVUtils
import java.io.File
import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

class MainActivity : Activity() {
    private val activityScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private lateinit var statusView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(createContentView())
        statusView.text = "Preparing official PP-OCR benchmark..."

        activityScope.launch {
            val message = try {
                withContext(Dispatchers.IO) {
                    runBenchmark()
                }
            } catch (error: Throwable) {
                if (error is CancellationException) throw error
                Log.e(TAG, "benchmark failed", error)
                "result=ERROR\n${error.javaClass.simpleName}: ${error.message}"
            }
            statusView.text = message
        }
    }

    override fun onDestroy() {
        activityScope.cancel()
        super.onDestroy()
    }

    private fun createContentView(): ScrollView {
        val scrollView = ScrollView(this)
        val layout = LinearLayout(this)
        layout.orientation = LinearLayout.VERTICAL
        layout.gravity = Gravity.START
        val padding = dp(20)
        layout.setPadding(padding, padding, padding, padding)

        val title = TextView(this)
        title.text = "Smartisax OCR Official Bench"
        title.textSize = 22f
        title.gravity = Gravity.START
        layout.addView(title)

        statusView = TextView(this)
        statusView.textSize = 14f
        statusView.setPadding(0, dp(14), 0, 0)
        statusView.setTextIsSelectable(true)
        layout.addView(statusView)

        scrollView.addView(layout)
        return scrollView
    }

    private fun dp(value: Int): Int = (value * resources.displayMetrics.density).toInt()

    private suspend fun runBenchmark(): String {
        val appRoot = getExternalFilesDir(null) ?: filesDir
        val image = resolvePath(
            intent.getStringExtra(EXTRA_IMAGE_PATH),
            File(File(appRoot, "input"), DEFAULT_IMAGE_NAME),
        )
        val result = resolvePath(
            intent.getStringExtra(EXTRA_RESULT_PATH),
            File(File(appRoot, "results"), "last-result.json"),
        )

        val payload = buildPredictionPayload(image)
        writeJson(result, payload)

        val message = "result=${payload.optString("result")}\n" +
            "image=${image.absolutePath}\n" +
            "output=${result.absolutePath}"
        Log.i(TAG, message)
        return message
    }

    private fun resolvePath(path: String?, fallback: File): File {
        if (path.isNullOrBlank()) return fallback
        val file = File(path.trim())
        if (file.isAbsolute) return file
        val appRoot = getExternalFilesDir(null) ?: filesDir
        return File(appRoot, path.trim())
    }

    private suspend fun buildPredictionPayload(image: File): JSONObject {
        val started = SystemClock.elapsedRealtime()
        val probe = probeImage(image)
        var status = probe.status
        var runResult: OCRRunResult? = null
        var runError: String? = null
        var coldLoadMs = 0L

        if (status == "IMAGE_READY") {
            var ocr: PaddleOCR? = null
            try {
                if (!OpenCVUtils.init(this)) {
                    throw IllegalStateException("Failed to initialize OpenCV native library")
                }
                ocr = PaddleOCR.create(
                    context = this,
                    config = OCR_CONFIG,
                    engineConfig = EngineConfig(numThreads = CPU_THREAD_NUM),
                    detModelAssetPath = DET_MODEL_ASSET,
                    recModelAssetPath = REC_MODEL_ASSET,
                    recConfigAssetPath = REC_CONFIG_ASSET,
                )
                coldLoadMs = ocr.coldLoadTimeMs
                runResult = ocr.recognize(image.readBytes())
                status = "OK"
            } catch (error: Throwable) {
                if (error is CancellationException) throw error
                status = "OCR_ERROR"
                runError = "${error.javaClass.simpleName}: ${error.message}"
                Log.e(TAG, "official OCR runtime failed", error)
            } finally {
                ocr?.release()
            }
        }

        val elapsedMs = SystemClock.elapsedRealtime() - started
        val sample = JSONObject()
            .put("id", imageId(image))
            .put("source_image", image.absolutePath)
            .put("status", status)
            .put("image_sha256", probe.sha256 ?: JSONObject.NULL)
            .put("image_size", imageSizeArray(probe.width, probe.height))
            .put("latency_ms", runResult?.totalTimeMs ?: elapsedMs)
            .put("wall_elapsed_ms", elapsedMs)
            .put("cold_load_ms", coldLoadMs)
            .put("peak_pss_kb", currentPssKb())
            .put("ppocr", ppocrLines(runResult))
            .put("native_metrics", nativeMetrics(runResult, coldLoadMs))

        probe.error?.let { sample.put("error", it) }
        runError?.let { sample.put("runtime_error", it) }

        val samples = JSONArray().put(sample)

        return JSONObject()
            .put("generated_at", utcNow())
            .put("kind", "textboom-ppocr-official-runtime-prediction")
            .put("boundary", "standalone official ppocr-sdk benchmark APK; no TextBoom, ROM, or data mutation")
            .put("engine", engineJson())
            .put("model", modelJson())
            .put("device", deviceJson())
            .put("samples", samples)
            .put("result", status)
    }

    private fun nativeMetrics(runResult: OCRRunResult?, coldLoadMs: Long): JSONObject {
        if (runResult == null) {
            return JSONObject()
                .put("cold_load_ms", coldLoadMs)
                .put("line_count", 0)
        }
        return JSONObject()
            .put("cold_load_ms", coldLoadMs)
            .put("det_ms", runResult.detectionTimeMs)
            .put("det_preprocess_ms", runResult.detPreprocessMs)
            .put("det_inference_ms", runResult.detInferenceMs)
            .put("det_postprocess_ms", runResult.detPostprocessMs)
            .put("rec_ms", runResult.recognitionTimeMs)
            .put("rec_preprocess_ms", runResult.recPreprocessMs)
            .put("rec_inference_ms", runResult.recInferenceMs)
            .put("rec_postprocess_ms", runResult.recPostprocessMs)
            .put("pipeline_overhead_ms", runResult.pipelineOverheadMs)
            .put("total_ms", runResult.totalTimeMs)
            .put("line_count", runResult.lineCount)
            .put("det_input_shape", intArrayJson(runResult.detInputShape))
            .put("rec_input_shapes", intArrayListJson(runResult.recInputShapes))
            .put("per_line_rec_ms", longArrayJson(runResult.perLineRecMs))
    }

    private fun ppocrLines(runResult: OCRRunResult?): JSONArray {
        val lines = JSONArray()
        runResult?.results?.forEachIndexed { index, result ->
            lines.put(
                JSONObject()
                    .put("index", index)
                    .put("text", result.text)
                    .put("confidence", result.confidence.toDouble())
                    .put("box", boxJson(result.box)),
            )
        }
        return lines
    }

    private fun boxJson(box: OCRBox): JSONObject {
        val points = JSONArray()
        var minX = Float.POSITIVE_INFINITY
        var minY = Float.POSITIVE_INFINITY
        var maxX = Float.NEGATIVE_INFINITY
        var maxY = Float.NEGATIVE_INFINITY
        box.points.forEach { point ->
            points.put(JSONArray().put(point.x.toDouble()).put(point.y.toDouble()))
            minX = minOf(minX, point.x)
            minY = minOf(minY, point.y)
            maxX = maxOf(maxX, point.x)
            maxY = maxOf(maxY, point.y)
        }
        return JSONObject()
            .put("points", points)
            .put("rect", JSONArray().put(minX.toDouble()).put(minY.toDouble()).put(maxX.toDouble()).put(maxY.toDouble()))
    }

    private fun engineJson(): JSONObject = JSONObject()
        .put("id", "official-paddleocr-ppocr-android")
        .put("version", "0.1.1")
        .put("runtime", "onnxruntime-android")
        .put("onnxruntime_android_version", ORT_ANDROID_VERSION)
        .put("opencv_version", OPENCV_VERSION)
        .put("cpu_threads", CPU_THREAD_NUM)
        .put("rec_batch_size", OCR_CONFIG.recBatchSize)

    private fun modelJson(): JSONObject = JSONObject()
        .put("id", MODEL_ID)
        .put("family", "PP-OCRv6 small")
        .put("det", DET_MODEL_ASSET)
        .put("rec", REC_MODEL_ASSET)
        .put("rec_config", REC_CONFIG_ASSET)
        .put("det_limit_side_len", OCR_CONFIG.detLimitSideLen)
        .put("det_limit_type", OCR_CONFIG.detLimitType)
        .put("det_thresh", OCR_CONFIG.detThresh.toDouble())
        .put("det_box_thresh", OCR_CONFIG.detBoxThresh.toDouble())
        .put("det_unclip_ratio", OCR_CONFIG.detUnclipRatio.toDouble())
        .put("rec_score_thresh", OCR_CONFIG.recScoreThresh.toDouble())

    private fun deviceJson(): JSONObject = JSONObject()
        .put("manufacturer", Build.MANUFACTURER)
        .put("brand", Build.BRAND)
        .put("model", Build.MODEL)
        .put("device", Build.DEVICE)
        .put("sdk_int", Build.VERSION.SDK_INT)
        .put("android_release", Build.VERSION.RELEASE)

    private fun probeImage(image: File): ImageProbe {
        if (!image.isFile) {
            return ImageProbe.error("IMAGE_MISSING", 0, 0, null, "missing file")
        }
        return try {
            val options = BitmapFactory.Options()
            options.inJustDecodeBounds = true
            BitmapFactory.decodeFile(image.absolutePath, options)
            if (options.outWidth <= 0 || options.outHeight <= 0) {
                ImageProbe.error("IMAGE_DECODE_FAILED", 0, 0, sha256(image), "BitmapFactory returned empty bounds")
            } else {
                ImageProbe("IMAGE_READY", options.outWidth, options.outHeight, sha256(image), null)
            }
        } catch (error: Throwable) {
            ImageProbe.error("IMAGE_PROBE_ERROR", 0, 0, null, "${error.javaClass.simpleName}: ${error.message}")
        }
    }

    private fun currentPssKb(): Int {
        val memoryInfo = Debug.MemoryInfo()
        Debug.getMemoryInfo(memoryInfo)
        return memoryInfo.totalPss
    }

    private fun imageSizeArray(width: Int, height: Int): JSONArray = JSONArray().put(width).put(height)

    private fun imageId(image: File): String {
        val name = image.name
        val dot = name.lastIndexOf('.')
        return when {
            dot > 0 -> name.substring(0, dot)
            name.isNotEmpty() -> name
            else -> "sample"
        }
    }

    private fun writeJson(result: File, payload: JSONObject) {
        val parent = result.parentFile
        if (parent != null && !parent.isDirectory && !parent.mkdirs()) {
            throw IllegalStateException("failed to create $parent")
        }
        result.writeText(payload.toString(2) + "\n", Charsets.UTF_8)
    }

    private fun sha256(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buffer = ByteArray(8192)
            while (true) {
                val read = input.read(buffer)
                if (read < 0) break
                digest.update(buffer, 0, read)
            }
        }
        return digest.digest().joinToString("") { "%02x".format(Locale.US, it.toInt() and 0xff) }
    }

    private fun utcNow(): String {
        val format = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        format.timeZone = TimeZone.getTimeZone("UTC")
        return format.format(Date())
    }

    private fun intArrayJson(values: List<Int>): JSONArray {
        val array = JSONArray()
        values.forEach { array.put(it) }
        return array
    }

    private fun longArrayJson(values: List<Long>): JSONArray {
        val array = JSONArray()
        values.forEach { array.put(it) }
        return array
    }

    private fun intArrayListJson(values: List<List<Int>>): JSONArray {
        val array = JSONArray()
        values.forEach { array.put(intArrayJson(it)) }
        return array
    }

    private data class ImageProbe(
        val status: String,
        val width: Int,
        val height: Int,
        val sha256: String?,
        val error: String?,
    ) {
        companion object {
            fun error(status: String, width: Int, height: Int, sha256: String?, error: String): ImageProbe {
                return ImageProbe(status, width, height, sha256, error)
            }
        }
    }

    companion object {
        private const val TAG = "SmartisaxOcrOfficialBench"
        private const val EXTRA_IMAGE_PATH = "image_path"
        private const val EXTRA_RESULT_PATH = "result_path"
        private const val DEFAULT_IMAGE_NAME = "imageboom.jpg"
        private const val MODEL_ID = "PP-OCRv6_small"
        private const val DET_MODEL_ASSET = "models/det/inference.onnx"
        private const val REC_MODEL_ASSET = "models/rec/inference.onnx"
        private const val REC_CONFIG_ASSET = "models/rec/inference.yml"
        private const val ORT_ANDROID_VERSION = "1.21.1"
        private const val OPENCV_VERSION = "4.9.0-official-aar"
        private const val CPU_THREAD_NUM = 4

        private val OCR_CONFIG = PaddleOCRConfig(
            recScoreThresh = 0.0f,
            recBatchSize = 1,
        )
    }
}
